rai/client.go

```diff
@@ -885,108 +885,123 @@ func (c *Client) ListOAuthClients() ([]OAuthClient, error) {
 		if item.Name == model {
 			return &item, nil
 		}
 	}
+
 	return nil, ErrNotFound
 }
 
 func (c *Client) LoadModel(
 	database, engine, name string, r io.Reader,
-) (*TransactionResult, error) {
+) (*TransactionAsyncResult, error) {
 	return c.LoadModels(database, engine, map[string]io.Reader{name: r})
 }
 
 func (c *Client) LoadModels(
 	database, engine string, models map[string]io.Reader,
-) (*TransactionResult, error) {
-	var result TransactionResult
-	tx := Transaction{
-		Region:   c.Region,
-		Database: database,
-		Engine:   engine,
-		Mode:     "OPEN",
-		Readonly: false}
-	actions := []DbAction{}
+) (*TransactionAsyncResult, error) {
+	queries := make([]string, 0)
 	for name, r := range models {
 		model, err := ioutil.ReadAll(r)
 		if err != nil {
 			return nil, err
 		}
-		action := makeLoadModelAction(name, string(model))
-		actions = append(actions, action)
+
+		queries = append(queries, fmt.Sprintf("def insert:rel:catalog:model[\"%s\"] = \"\"\"%s\"\"\"", name, model))
 	}
-	data := tx.Payload(actions...)
-	err := c.Post(PathTransaction, tx.QueryArgs(), data, &result)
-	if err != nil {
-		return nil, err
+
+	return c.Execute(database, engine, strings.Join(queries, "\n"), nil, false)
+}
+
+func (c *Client) LoadModelsAsync(
+	database, engine string, models map[string]io.Reader,
+) (*TransactionAsyncResult, error) {
+	queries := make([]string, 0)
+	for name, r := range models {
+		model, err := ioutil.ReadAll(r)
+		if err != nil {
+			return nil, err
+		}
+
+		queries = append(queries, fmt.Sprintf("def insert:rel:catalog:model[\"%s\"] = \"\"\"%s\"\"\"", name, model))
 	}
-	return &result, nil
+
+	return c.ExecuteAsync(database, engine, strings.Join(queries, "\n"), nil, false)
 }
 
 // Returns a list of model names for the given database.
 func (c *Client) ListModelNames(database, engine string) ([]string, error) {
-	var models listModelsResponse
-	tx := NewTransaction(c.Region, database, engine, "OPEN")
-	data := tx.Payload(makeListModelsAction())
-	err := c.Post(PathTransaction, tx.QueryArgs(), data, &models)
+	modelNames := make([]string, 0)
+	models, err := c.ListModels(database, engine)
 	if err != nil {
-		return nil, err
+		return modelNames, err
 	}
-	actions := models.Actions
-	// assert len(actions) == 1
-	result := []string{}
-	for _, model := range actions[0].Result.Models {
-		result = append(result, model.Name)
+
+	for _, model := range models {
+		modelNames = append(modelNames, model.Name)
 	}
-	return result, nil
+
+	return modelNames, nil
 }
 
 // Returns the names of models installed in the given database.
 func (c *Client) ListModels(database, engine string) ([]Model, error) {
-	var models listModelsResponse
-	tx := NewTransaction(c.Region, database, engine, "OPEN")
-	data := tx.Payload(makeListModelsAction())
-	err := c.Post(PathTransaction, tx.QueryArgs(), data, &models)
+	models := make([]Model, 0)
+	resp, err := c.Execute(database, engine, "def output:__models__ = rel:catalog:model", nil, true)
```

Comment:
again, why are we qualifying the output relation here?

Reply:
oh this is an old commit, we no longer query for all models (including name and source)

Reply:
we qualify output relation to prevent accidental collisions between output results